### ISSUE TITLE
Inconsistency in `invalid-return-type` With a `global` Call

### ISSUE THREAD

**Issue by krishsai7** [astral-sh/ty#4469]
### Summary

Sample code I was playing with:
```python
var1 = None
var2 = None

def f1(text: str) -> str:
    global var1, var2
    if var1 != text:
        var1 = text
        var2 = text.lower()

    return var2
```

The code above returns an error as I expect:
```
Return type does not match returned value: expected `str`, found `str | None` (invalid-return-type) [Ln 10, Col 12]

    Expected `str` because of return type 
    info: element `None` of union `str | None` is not assignable to `str`
```

If I remove line 7 (`var1 = text`), the error goes away. Using `pyrefly` as a reference, [I cannot reproduce the above](https://pyrefly.org/sandbox/?project=v2.pZRNDgIhDIWvUjMLXE0DSxPv4AFmg0rQhIApOHp8y48_xDgbV4SGvpB-r-9XZM6aZPPa5PmiXpdmUrlO5p7KwhXKfLbhWBf22rFvSOZdIvVhUBZdbaF0vn3U5HN1dOFmqGLuPEjqe94D7GoB2tc5aRiUvXKeMYBx8gNHmYFTSpe4QXy2B7JoPB7DIWLXgZCtpWd9LinZy0Eo_ogs-w9oUTmLJcwiU-YXi-Qe).

[Playground](https://play.ty.dev/0c2a7131-456f-45b4-946a-80104a47a94a)

I did find that #311 seems related to this but that was marked Completed.

### Version

e014dc4df

**Comment by jelle-openai** [comment]
To make the report a bit more explicit (I was confused reading it), the bug here is that ty does not error on this code:

```python
var1 = None
var2 = None

def f1(text: str) -> str:
    global var1, var2
    if var1 != text:
        var2 = text.lower()

    return var2
```

**Comment by sharkdp** [comment]
Thank you for reporting this.

I understand that this is a bit surprising, but this is ty's intended behavior (see our reasoning [here](https://github.com/astral-sh/ty/issues/1822#issuecomment-4226251945)). We do infer the type of the global `var1` as `None`, based on the single `var1 = None` assignment in the global scope. Under that assumption, since `None` is always different from a `str`, ty understands that the `var1 != text` condition is always true. And if that `if` branch is always taken, `var2` will always have a `str` value, and the `return var2` is accepted.

If external mutation of that variable is expected, we do require an explicit annotation (e.g. `var1: str | None = None`).

Note: if you enable ty's new `redundant-condition-strict` rule, this would be much easier to understand. Compare: https://play.ty.dev/b2bb2d4a-31dd-4b9a-8205-c8775de38f39 (FYI @AlexWaygood).

**Comment by krishsai7** [comment]
> Under that assumption, since `None` is always different from a `str`, ty understands that the `var1 != text` condition is always true. And if that `if` branch is always taken, `var2` will always have a `str` value, and the `return var2` is accepted.

If this is true, why does the original code snippet raise an error? I'm working under the assumption that `ty` is able to shortcut branch logic like you've listed above.

First ever invocation: `var1` is seen as None, which is different from `text` and therefore `var1` is globally set to `text` and thus, `var2` is globally set to `text.lower()`. This returns a `str`.
Second invocation: `var2` is now set as a `str` and whether or not the branch fires is irrelevant. Since it was set on the previous invocation, that original `text.lower()` value must be returned (or the new `text.lower()` if the branch fires)

Toying with a snippet like this:
```python
var1 = None
var2 = None

def f1(text: str) -> str:
    global var1, var2
    if var1 != text:
        var1 = text
        var2 = text.lower()

    return var2

print(f1("test"))
print(f1(var1))
```

Is this just an issue with the original vars being set to `None`/as you mentioned, being unsure of if the variable was mutated outside of the method?

> `redundant-condition-strict`

This is neat, thank you for this + the example.

**Comment by sharkdp** [comment]
> First ever invocation: `var1` is seen as None, which is different from `text` and therefore `var1` is globally set to `text` and thus, `var2` is globally set to `text.lower()`. This returns a `str`. Second invocation: `var2` is now set as a `str` and whether or not the branch fires is irrelevant. Since it was set on the previous invocation, that original `text.lower()` value must be returned (or the new `text.lower()` if the branch fires)

That reasoning is correct, but ty's analysis is not *that* sophisticated. It can't toggle state between what happens in the first vs later iterations. Instead, it infers *one* type for `var1` that is compatible with all definition sites, and that type is `str | None` (which then leads to the downstream error).
